### PR TITLE
WT-5215 Add basic randomized test for doing incremental backups.

### PR DESCRIPTION
**Pull request by ddanderson**
We use various combinations of full and incremental backup and checkpoints.
Schema changes include creation, drop and rename.
Changes to file data include inserts, updates, modify and remove.

### CODE REVIEW

**Comment by ddanderson** [comment]
@sueloverso, this is almost done, I think it would be good to review.  This is a pretty basic randomized test at this point.  Unfortunately, it doesn't really accomplish the specific goals in the ticket.  We should discuss if it belongs on a different ticket or whether we should extend this one to satisfy some or all of the goals.

**Comment by ddanderson** [comment]
retest this please

File: test/csuite/incr_backup/main.c [hunk 1]
```diff
@@ -0,0 +1,846 @@
...
+static void usage(void) WT_GCC_FUNC_DECL_ATTRIBUTE((noreturn));
+
+/*
+ * Note: set this to true to copy incremental files completely.
+ */
+static bool slow_incremental = false;
+
+/* TODO: rename and drop are not currently working, they give resource busy. */
+static bool do_rename = false;
+static bool do_drop = false;
```
Comment (sueloverso) [inline test/csuite/incr_backup/main.c]: Ravi saw something like this when he tried to do some testing with the example program. He was also seeing EBUSY when I asked him why he was closing and reopening the connection entirely. Is it due to the table being left open or cache cursors on that table or something? If you're sure all cursor to the table are closed, we should investigate it.

File: test/csuite/incr_backup/main.c [hunk 2]
```diff
@@ -0,0 +1,846 @@
...
+      "**** FAILURE\n"
+      "To reproduce, please rerun with: %s -S %" PRIu64 "\n",
+      progname, seed);
+}
+
+/*
+ * key_value --
+ *     Return the key, value and operation type for a given change to a table. The first 10000
+ *     changes to a table are all inserts, the next 10000 are updates of the same records, the next
+ *     10000 are all modifications of the existing records, the next 10000 will be drops. Then we
```
Comment (sueloverso) [inline test/csuite/incr_backup/main.c]: I think `drops` should be `removes`.
Overall I really like this concept to know what we expect.

File: test/csuite/incr_backup/main.c [hunk 3]
```diff
@@ -0,0 +1,846 @@
...
+ *     repeat the cycle. That makes it easy on the checking side (knowing how many total changes
+ *     have been made) to check the state of the table.
+ *
+ * The keys generated are unique among the 10000, but we purposely don't make them sequential.
```
Comment (sueloverso) [inline test/csuite/incr_backup/main.c]: I would add more about why we don't make them sequential. To scatter the keys among the pages. (I think that's the reason.)

File: test/csuite/incr_backup/main.c [hunk 4]
```diff
@@ -0,0 +1,846 @@
...
+key_value(uint64_t change_count, char *key, size_t key_size, WT_ITEM *item, OPERATION_TYPE *typep)
+{
+    uint32_t key_num;
+    OPERATION_TYPE op_type;
+    size_t pos, value_size;
+    char *cp;
+    char ch;
+
+    key_num = change_count % 10000;
+    *typep = op_type = (OPERATION_TYPE)((change_count % 40000) / 10000);
```
Comment (sueloverso) [inline test/csuite/incr_backup/main.c]: I suggest making these #defines so if we decide 5000 or 80000 is right we only change it once. Is the 40000 just 10000 time the number of different operation types we have? Can that be computed?

File: test/csuite/incr_backup/main.c [hunk 5]
```diff
@@ -0,0 +1,846 @@
...
+    if (op_type == REMOVE)
+        return; /* remove needs no key */
+
+    /* The value sizes vary "predictably" between 10 and the max value size for this table. */
+    value_size = 10 + (change_count * 103) % (item->size - 20);
+    testutil_assert(item->size > value_size);
+
+    /*
+     * For a given key, a value is first inserted, then later updated, then modified. When a value
+     * is inserted, it is all the letter 'a'. When the value is updated, is mostly 'b', with some
```
Comment (sueloverso) [inline test/csuite/incr_backup/main.c]: 'is mostly...' should be 'it is mostly'

File: test/csuite/incr_backup/main.c [hunk 6]
```diff
@@ -0,0 +1,846 @@
...
+     * 'c' mixed in. When the value is to modified, we'll end up with a value with mostly 'b' and
```
Comment (sueloverso) [inline test/csuite/incr_backup/main.c]: 'value is to modified': remove 'to'

File: test/csuite/incr_backup/main.c [hunk 7]
```diff
@@ -0,0 +1,846 @@
...
+     * 'M' mixed in in different spots. Thus the modify operation will have both additions ('M') and
```
Comment (sueloverso) [inline test/csuite/incr_backup/main.c]: add a comma between the two 'in's

File: test/csuite/incr_backup/main.c [hunk 8]
```diff
@@ -0,0 +1,846 @@
...
+    *dest = *src;
+    WT_CLEAR(*src);
+}
+
+/*
+ * table_updates --
+ *     Potentially make changes to a single table.
+ */
+static void
+table_updates(WT_SESSION *session, TABLE *table)
```
Comment (sueloverso) [inline test/csuite/incr_backup/main.c]: Can you consider a slightly different name for the function? I thought it referred specifically to `cursor->update`. Maybe `table_mods` or `table_changes` or `do_work` ...?

File: test/csuite/incr_backup/main.c [hunk 9]
```diff
@@ -0,0 +1,846 @@
...
+    uint32_t count;
+} ACTIVE_FILES;
+
+extern int __wt_optind;
+extern char *__wt_optarg;
+
+/*
+ * The choices of operations we do to each table.
+ */
+typedef enum { INSERT, UPDATE, MODIFY, REMOVE } OPERATION_TYPE;
```
Comment (sueloverso) [inline test/csuite/incr_backup/main.c]: Alphabetize these.

File: test/csuite/incr_backup/main.c [hunk 10]
```diff
@@ -0,0 +1,846 @@
...
+            item.data = value;
+            item.size = table->max_value_size;
+            key_value(change_count, key, sizeof(key), &item, &op_type);
+            cur->set_key(cur, key);
+            switch (op_type) {
+            case INSERT:
+                cur->set_value(cur, &item);
+                testutil_check(cur->insert(cur));
+                break;
+            case UPDATE:
```
Comment (sueloverso) [inline test/csuite/incr_backup/main.c]: Alphabetize the cases too.

File: test/csuite/incr_backup/main.c [hunk 11]
```diff
@@ -0,0 +1,846 @@
...
+    op_type = (OPERATION_TYPE)(total_changes % 40000) / 10000;
+    value = dcalloc(1, table->max_value_size);
+
+    VERBOSE(3, "Checking: %s\n", table->name);
+    switch (op_type) {
+    case INSERT:
+        expect_records = total_changes % 10000;
+        break;
+    case UPDATE:
+    case MODIFY:
```
Comment (sueloverso) [inline test/csuite/incr_backup/main.c]: Alphabetize here too

File: test/csuite/incr_backup/main.c [hunk 12]
```diff
@@ -0,0 +1,846 @@
...
+    char key[MAX_KEY_SIZE];
+
+    /*
+     * We change each table in use about half the time.
+     */
+    if (__wt_random(&table->rand) % 2 == 0) {
+        value = dcalloc(1, table->max_value_size);
+        value2 = dcalloc(1, table->max_value_size);
+        nrecords = __wt_random(&table->rand) % 1000;
+        VERBOSE(4, "inserting %d records into %s\n", (int)nrecords, table->name);
```
Comment (sueloverso) [inline test/csuite/incr_backup/main.c]: We're not only inserting. So maybe "changing %d record in %s" or "operating on %d records in %s"?

File: test/csuite/incr_backup/main.c [hunk 13]
```diff
@@ -0,0 +1,846 @@
...
+
+static void
+drop_table(WT_SESSION *session, TABLE_INFO *tinfo, uint32_t slot)
+{
+    char *uri;
+
+    testutil_assert(TABLE_VALID(&tinfo->table[slot]));
+    uri = tinfo->table[slot].name;
+
+    VERBOSE(3, "create %s\n", uri);
```
Comment (sueloverso) [inline test/csuite/incr_backup/main.c]: 'create' should be 'drop'.

File: test/csuite/incr_backup/main.c [hunk 14]
```diff
@@ -0,0 +1,846 @@
...
+    }
+
+    testutil_check(session->open_cursor(session, table->name, NULL, NULL, &cursor));
+    got_records = 0;
+    while ((ret = cursor->next(cursor)) == 0) {
+        got_records++;
+        testutil_check(cursor->get_key(cursor, &got_key));
+        testutil_check(cursor->get_value(cursor, &got_value));
+        testutil_assert(sscanf(got_key, KEY_FORMAT, &keylow, &keyhigh) == 2);
+        change_count = (u_int)keyhigh * 100 + (u_int)keylow;
```
Comment (sueloverso) [inline test/csuite/incr_backup/main.c]: A brief comment reminding us about how we're doing this checking here would be helpful.

File: test/csuite/incr_backup/main.c [hunk 15]
```diff
@@ -0,0 +1,846 @@
...
+
+    /* Move the current active list to the master list */
+    active_files_move(master_active, &active);
+}
+
+/*
+ * Verify the backup to make sure the proper tables exist and have the correct content.
+ */
+static void
+check_backup(const char *backup_home, const char *backup_check, TABLE_INFO *tinfo)
```
Comment (sueloverso) [inline test/csuite/incr_backup/main.c]: I suggest moving the `check_table` function next to this function.

File: test/csuite/incr_backup/main.c [hunk 16]
```diff
@@ -0,0 +1,846 @@
...
+    for (iter = 0; iter < ITERATIONS; iter++) {
+        VERBOSE(1, "**** iteration %d ****\n", (int)iter);
+
+        /*
+         * We have schema changes during about half the iterations. The number of schema changes
+         * varies, averaging 10.
+         */
+        if (tinfo.tables_in_use == 0 || __wt_random(&rnd) % 2 != 0) {
+            while (__wt_random(&rnd) % 10 != 0) {
+
```
Comment (sueloverso) [inline test/csuite/incr_backup/main.c]: Remove this blank line.

File: test/csuite/incr_backup/main.c [hunk 17]
```diff
@@ -0,0 +1,846 @@
...
+                else if (__wt_random(&rnd) % 3 == 0 && do_rename)
+                    rename_table(session, &tinfo, slot);
+                else if (do_drop)
+                    drop_table(session, &tinfo, slot);
+            }
+        }
+        for (slot = 0; slot < tinfo.table_count; slot++) {
+            if (TABLE_VALID(&tinfo.table[slot])) {
+                table_updates(session, &tinfo.table[slot]);
+            }
```
Comment (sueloverso) [inline test/csuite/incr_backup/main.c]: Remove extra curly braces.

File: test/csuite/incr_backup/main.c (hunk 1 above)
Comment (ddanderson) [reply test/csuite/incr_backup/main.c]: rename and drop are returning `EBUSY`.  I don't think I've left cursors open, I can investigate further.

Comment (sueloverso) [reply test/csuite/incr_backup/main.c]: Don't take too much time on this right now. I'd rather get this committed and open up follow on work for after the release.

File: test/csuite/incr_backup/main.c [hunk 18]
```diff
@@ -0,0 +1,869 @@
...
+    /* Compute a random value of file_max. */
+    rough_size = __wt_random(&rnd) % 3;
+    if (rough_size == 0)
+        file_max = 100 + __wt_random(&rnd) % 100; /* small log files, min 100K */
+    else if (rough_size == 1)
+        file_max = 200 + __wt_random(&rnd) % 1000; /* 200K to ~1M */
+    else
+        file_max = 1000 + __wt_random(&rnd) % 20000; /* 1M to ~20M */
+    testutil_check(__wt_snprintf(
+      conf, sizeof(conf), "create,log=(enabled=true,file_max=%" PRIu32 "K)", file_max));
```
Comment (sueloverso) [inline test/csuite/incr_backup/main.c]: One suggestion is that at some verbose level you might also add `verbose=(backup)` to the config. I just did this exact thing to debug the CREATE issue.

**Pull request merged**